### ISSUE TITLE
clang should diagnose `memcpy()` via `-Wunsafe_buffer_usage`

### ISSUE THREAD

**Issue by pkasting** [llvm/llvm-project#119405]
When `-Wunsafe-buffer-usage` is on, `memcpy()`  should trigger a warning (unless locally `pragma`d away), since it effectively does pointer math under the hood (takes a ptr and size).

Similar comments would apply to e.g. `memset()`, `memmove()`, etc.

**Comment by nickdesaulniers** [comment]
Sorry, trying to parse the description (thanks for filing a bug to track this either way). Are you suggesting we enable `-Wunsafe-buffer-usage` when building llvm-libc? (then proceed to fix any new instances of the warning that crop up)? Is this a new-ish compiler diagnostic? Does it exist in both clang and gcc?

**Comment by pkasting** [comment]
No; I'm suggesting that authors who call memcpy() in their own code should receive warnings if they compile said code with -Wunsafe-buffer-usage. I don't have any recommendation re: how libc itself should be compiled.

**Comment by nickdesaulniers** [comment]
To answer my own questions:
> Is this a new-ish compiler diagnostic?

clang-16+

> Does it exist in both clang and gcc?

Does not exist in gcc, yet.

> I'm suggesting that authors who call memcpy() in their own code should receive warnings if they compile said code with -Wunsafe-buffer-usage. I don't have any recommendation re: how libc itself should be compiled.

Ah I see! This issue was accidentally re-tagged as libc; this is a feature request against Clang! Restoring the initial tag which was correct. Sorry for my confusion.

**Comment by nickdesaulniers** [comment]
cc @ziqingluo-90

**Comment by nickdesaulniers** [comment]
Perhaps with a relevant godbolt link. 😉

**Comment by pkasting** [comment]
Was that implemented recently? I had filed this because back in early December Chromium wasn't seeing a warning in that case.

If so, this can be closed. If not, I'm a little mystified why we weren't seeing the warning, and I should try to dig deeper.

**Comment by nickdesaulniers** [comment]
> Was that implemented recently?

Yes.  In fact, you can find that answer yourself in godbolt by changing the compiler version in the dropdown.

It's implemented in clang trunk which is not yet released. The latest release version of clang-19 doesn't support this.

I don't think Chromium sticks to the upstream llvm release convention (it's in fact irrelevant to most LLVM teams at Google), but it's possible that when you checked last you had just missed this landing upstream.

I'll close this issue out, but I do appreciate you taking the time to report this.
